all/bodies/vision.py

```diff
@@ -9,23 +9,55 @@ def __init__(self, agent, size=4, lazy=False):
         self._frames = []
         self._size = size
         self._lazy = lazy
+        self._to_cache = ToCache()
 
     def process_state(self, state):
         if not self._frames:
             self._frames = [state.observation] * self._size
         else:
             self._frames = self._frames[1:] + [state.observation]
         if self._lazy:
-            return LazyState.from_state(state, self._frames)
+            return LazyState.from_state(state, self._frames, self._to_cache)
         if isinstance(state, StateArray):
             return state.update('observation', torch.cat(self._frames, dim=1))
         return state.update('observation', torch.cat(self._frames, dim=0))
 
 
+class ToCache:
+    def __init__(self, from_device=None, to_device=None, max_size=16):
+        self.from_device = from_device
+        self.to_device = to_device
+        self.max_size = max_size
+        self.cache_data = []
+
+    def convert(self, value, device):
+        if self.from_device is None:
+            self.from_device = value.device
+        if self.to_device is None:
+            self.to_device = device
+        if self.from_device != value.device or self.to_device != device:
```

Comment:
I find the usage a little confusing... `device` only has one valid value if `to_device` is set, but can be set to anything if `to_device` is none. Is there any reason not to choose one usage or the other?

Reply:
Ok, I greatly simplified this logic.